gh-137609: Update signatures of builtins in the documentation (#137610)
python/cpython: Doc/library/functions.rst

.. |func-bytearray| replace:: ``bytearray()``
.. |func-bytes| replace:: ``bytes()``

.. function:: abs(number, /)

>    Return the absolute value of a number.  The argument may be an

Reviewer: ```suggestion
   Return the absolute value of *number*.  *number* may be an
```

Author: I think that there is no need to refer the argument by name if it is the only argument. I think it's easier to understand if you don't get distracted by the name. And if you were to read it out loud, which variant sounds better?

          return False

.. function:: ascii(object, /)

   As :func:`repr`, return a string containing a printable representation of an
   object, but escape the non-ASCII characters in the string returned by
   :func:`repr` using ``\x``, ``\u``, or ``\U`` escapes.  This generates a string
   similar to that returned by :func:`repr` in Python 2.

> .. function:: bin(integer, /)

Reviewer 2: 'integer' could be too close to `int` for confusion
```suggestion
.. function:: bin(num, /)
```

Author: I initially used `number` (it is used in the builtin signature), but changed it to integer, since only `int` and objects implementing `__index__` are supported. In many other parts of the documentation "integer" means "`int` or objects implementing `__index__`".

   .. versionadded:: 3.7

.. _func-bytearray:
.. class:: bytearray()
           bytearray(source)
           bytearray(source, encoding)

>            bytearray(source, encoding, errors)

Reviewer 3: I believe only two lines are needed.
```suggestion
.. class:: bytearray(source=b'')
           bytearray(source, encoding, errors='strict')
```
See also #137100, which is also about the text that follows.

Author: Some descriptions have separate signature for no argument, others merge it with a signature with one argument. See for example `dict` which could be written as `dict(mapping_or_iterable=(), **kwargs)`, but is written as three semantically different signatures. I tried to be more consistent and chose the former variant. But I have no such strong preference.

Reviewer 3: I prefer fewer lines, and approve of the changes.

Reviewer 2: I'm not sure I agree here, `source=b''` makes it less clear to me that `source` also accepts e.g. iterables of integers, buffer-protocol, etc. I would suggest:

```rst
.. class:: bytearray()
           bytearray(source, /)
           bytearray(source, /, encoding, errors='strict')
```

Note I have suggested to annotate 'source' as positional-only -- I think this makes more sense to users than writing e.g. `bytearray(source=my_numpy_array)`. My IDE & type-checkers also indicates that `source=...` is an error, as it is annotated as positional-only [in typeshed](https://github.com/python/typeshed/blob/main/stdlib/builtins.pyi#L748-L754).

.. _func-bytes:
.. class:: bytes()

>            bytes(source)

Reviewer 3: Same comments as for bytearray.

      in Python 3.2.

> .. function:: chr(i, /)

Reviewer 2: ```suggestion
.. function:: chr(codepoint, /)
```

Author: I thought on changing this name, but then we need to change also the following text and the docstring "Return a Unicode string of one character with ordinal i; 0 <= i <= 0x10ffff."

.. _func-frozenset:

> .. class:: frozenset(iterable=(), /)

Reviewer 2: The empty frozenset is (notionally) a singleton, similar to the empty tuple, so I think it's clearer to distinguish them here:
```suggestion
.. class:: frozenset()
           frozenset(iterable, /)
```

Reviewer 4: Either approach could work but I like the way it's in the current PR; it's better to have fewer distinct signatures.

      *classinfo* can be a :ref:`types-union`.

.. function:: iter(iterable, /)
              iter(callable, sentinel, /)

   Return an :term:`iterator` object.  The first argument is interpreted very
   differently depending on the presence of the second argument. Without a

>    second argument, the sinle argument must be a collection object which supports the

Reviewer 3: ```suggestion
   second argument, the single argument must be a collection object which supports the
```

.. _func-list:

> .. class:: list(iterable=(), /)

Reviewer 2: As with frozenset I think clearer to keep two lines here:

```suggestion
.. class:: list()
           list(iterable, /)
```

Author: Your suggestions are opposite to @terryjreedy's. I just changed the code in opposite direction.

Reviewer 2: @terryjreedy what do you think? I would be -1 on combining here, I think it goes too far (even though it is technically accurate).

We currently have two lines for these container initialisers in the documentation, so it's not adding more bloat.

      Added the *strict* parameter.

.. function:: max(iterable, /, *, key=None)

>               max(iterable, /, *, default, key=None)

Reviewer 3: Why the inconsistency of type mode, normal versus italic?  Applies to next line also.

Author: I guess this is a flaw of the GitHub highligher.

>               max(arg1, arg2, /, *args, key=None)

Reviewer 3: To me, the `/,` belongs after `*args` as the latter only collects the remaining positional args.

Author: This is not valid syntax.

Reviewer 3: Right: `SyntaxError: / must be ahead of *`.  `*args` has meaning of `*` + 'collect remaining positionals'.

   :ref:`typememoryview` for more information.

.. function:: min(iterable, /, *, key=None)
              min(iterable, /, *, default, key=None)

>               min(arg1, arg2, /, *args, key=None)

Reviewer 5: See https://github.com/python/cpython/pull/131868 for min/max.

Reviewer 2: Merged

   .. versionchanged:: 3.11
      The ``'U'`` mode has been removed.

.. function:: ord(character, /)

>    The argument must be a one-character string or a :class:`bytes` or

Author: I created a separate issue for this: #137668.

   .. versionchanged:: 3.11
      The ``'U'`` mode has been removed.

> .. function:: ord(c, /)

Reviewer 2: ```suggestion
.. function:: ord(chr, /)
```

Author: Hmm, the following text and the docstring are incorrect, they should be changed in any case.

.. _func-range:
.. class:: range(stop, /)

>            range(start, stop, step=1, /)

Reviewer 5: See https://github.com/python/cpython/pull/125945 for range

Reviewer 3: That PR, submitted last October, proposed these same changes.  There was some discussion about adding the `/`s, perhaps in part because only in one place, but there are other approvals.  It should be merged before this one so that the OP, a new contributor, gets proper credit.

Reviewer 2: Merged

            return f"Person('{self.name}', {self.age})"

.. function:: reversed(sequence, /)

>    Return a reverse :term:`iterator`.  The argument must be an object which has

Reviewer: Why not update the previous?

```suggestion
   Return a reverse :term:`iterator`.  *sequence* must be an object which has
```

Author: Actually, `sequence` is not so good name. The argument can be a dict.

.. _func-set:

> .. class:: set(iterable=(), /)

Reviewer 2: As with frozenset & list:
```suggestion
.. class:: set()
           set(iterable, /)
```

   single: string; str() (built-in function)

.. _func-str:
.. class:: str()
           str(object)
           str(object, encoding)
           str(object, encoding, errors)

>            str(object, *, errors)

Reviewer 3: This is tough.  There is a default encoding if `errors` is given, not if not.  `str(b'abc')=="b'abc'"` (the 6 char repr()) while `str(b'abc',errors='strict')=='abc'` (3 chars).

   single: string; str() (built-in function)

.. _func-str:

> .. class:: str(*, encoding='utf-8', errors='strict')

Reviewer 4: It's technically true that you can pass just the encoding or errors kwargs (like `str(encoding='ebcdic')`), but it's not practically useful, since you just get an empty string whatever encoding you pass, and I fear it's confusing to users to put this here.

The old docs here feel clearer even if they're not as precise.

Author: We should be technically correct.

Several years ago I proposed to deprecate this case, but my proposition was rejected. If it is worth to continue supporting this in code, it is worth to document it. Note that values of encoding and errors are ignored:
```pycon
>>> str(encoding='spam', errors='ham')
''
```

.. _func-tuple:

> .. class:: tuple(iterable=(), /)

Reviewer 2: As with the other collection types. This one uses an empty tuple to initialise an empty tuple!
```suggestion
.. class:: tuple()
           tuple(iterable, /)
```

   :noindex:

   Rather than being a function, :class:`tuple` is actually an immutable
   sequence type, as documented in :ref:`typesseq-tuple` and :ref:`typesseq`.

.. class:: type(object, /)

>            type(name, bases, dict, /, **kwds)

Reviewer 2: We use `**kwargs` elsewhere

```suggestion
           type(name, bases, dict, /, **kwargs)
```

Author: I think that historically `kwds` was used first, but now `kwargs` is used 4-5 times more. `kwds` is still used 93 times in the documentation and 519 in the Python code. It is not error, but we can change the name here.
